File: ptp-prolog/src/SectionIB/task65.pl
:- module(section_ib_task65, [split_prime_nonprime/3]).

split_prime_nonprime([], [], []).
split_prime_nonprime([Head|Tail], [Head|Primes], NonPrimes) :-
    is_prime(Head),
    split_prime_nonprime(Tail, Primes, NonPrimes).
split_prime_nonprime([Head|Tail], Primes, [Head|NonPrimes]) :-
    \+ is_prime(Head),
    split_prime_nonprime(Tail, Primes, NonPrimes).

is_prime(2).
is_prime(Number) :-
    integer(Number),
    Number > 2,
    Number mod 2 =\= 0,
    has_no_odd_divisor_from(Number, 3).

has_no_odd_divisor_from(Number, Divisor) :-
    Divisor * Divisor > Number.
has_no_odd_divisor_from(Number, Divisor) :-
    Number mod Divisor =\= 0,
    NextDivisor is Divisor + 2,
    has_no_odd_divisor_from(Number, NextDivisor).

File: ptp-haskell/src/SectionIB/Task65.hs
module SectionIB.Task65
  ( splitPrimeNonPrime
  ) where

import Data.List (partition)

isPrime :: Int -> Bool
isPrime n
  | n <= 1 = False
  | n == 2 = True
  | even n = False
  | otherwise = hasNoOddDivisorFrom 3
  where
    hasNoOddDivisorFrom :: Int -> Bool
    hasNoOddDivisorFrom divisor
      | divisor * divisor > n = True
      | n `mod` divisor == 0 = False
      | otherwise = hasNoOddDivisorFrom (divisor + 2)

splitPrimeNonPrime :: [Int] -> ([Int], [Int])
splitPrimeNonPrime = partition isPrime
